pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -2333,7 +2344,7 @@ private void batchMessageAndSend(boolean shouldScheduleNextBatchFlush) {
         }
     }
 
-    protected void processOpSendMsg(OpSendMsg op) {
+    protected synchronized void processOpSendMsg(OpSendMsg op) {
```

Comment:
Why do we need this change? It looks like there no connection state check in this method.

Reply:
Since the collection `pendingMessages` is not thread-safety, I want to protect it in the current PR.

Reply:
I think this makes sense and prefer this style. This doesn't actually change the current behavior since processOpSendMsg is already called in all locations from other synchronized methods or `synchronized` code blocks. However, I personally prefer making synchronized methods explicit although the current style of the Pulsar code base is mixed. 
It seems that there has been some assumption that it would be less expensive to not put `synchronized` on methods. I don't think that there's a performance cost when methods are `synchronized` and when the thread is already owning the lock.